Update integration tests to store id of erroring policies

khieta opened the pull request:
## Description of changes

This PR updates the “errors” field of the integration tests to store only the id of erroring policies. This will prevent us from having to change the corpus tests whenever an error message changes (which is so far the only reason we’ve had to update the tests), and going forward (#586) will allow for easier comparison with other implementations of Cedar that may produce different messages, but should agree on which policies error.

Specifically, this PR:
* Adds `id` method for `AuthorizationError` (requires a minor version bump)
* Updates integration test format
    * Exposes `JsonTest` and `JsonRequest` fields for reuse in `cedar-spec` and `cedar-policy-cli`
    * Changes errors field to list policy ids rather than error messages
    * Updates corpus tests to match new test format. You can skip this part of the diff.

## Issue #, if available

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [X] A backwards-compatible change requiring a minor version bump to `cedar-policy` (e.g., addition of a new API).

I confirm that this PR (choose one, and delete the other options):

- [X] Updates the "Unreleased" section of the CHANGELOG with a description of my change.

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [X] equires updates, and I have made / will make these updates myself.

## Disclaimer

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

khieta commented:
Apologies for the absurdly large diff. The important part of the diff (i.e., the code outside `cedar-integration-tests/corpus_tests`) is small!

khieta commented:
Here are a few things I noticed, but didn’t change because they would require a major version bump. Lmk if you think these should be fixed in a followup PR, or if i'm just being picky!

1. The “errors” field of the `InterfaceDiagnostics` type should by `HashSet<PolicyId>` instead of `HashSet<String>`.
2. In the `Response` structure, the field is called `reason` (singular), whereas in the integration tests it is called `reasons` (plural). I didn’t change these to be consistent because I’m conflicted about the right option. I think “reasons” makes more sense since there may be multiple policy ids (and it’s consistent with “errors”), but changing the `Response` type is a breaking change.
3. Rather than adding an `error_ids` method to `Diagnostics`, it would make more sense to add an `id` method to the publicly exposed `AuthorizationError` type that returns a `PolicyId`. I don’t think this is possible without a breaking change.

khieta commented:
Looks like there's a downstream failure in CedarJava. This should be easy to fix, but does beg the question of whether this is PR is truly backwards compatible. I maintain that it is because our integration test format should not be considered part of the stable API... but open to second opinions.

john-h-kastner-aws commented on cedar-policy-core/src/authorizer/err.rs:
```diff
@@ -32,3 +32,12 @@ pub enum AuthorizationError {
+
+impl AuthorizationError {
+    /// Get the id of the erroring policy
+    pub fn id(&self) -> &PolicyID {
```
Should this not  be `pub` since `AuthorizationError` is `pub use`'ed but this return the internal `PolicyId` type?

The enum already exposes this to anyone who matches on it, but returning it from a function could be more confusing IMO. Better to just have the function on `Diagnostics` that returns the wrapped `PolicyId` type.

john-h-kastner-aws replied:
Responding to your earlier comment: 

> Rather than adding an error_ids method to Diagnostics, it would make more sense to add an id method to the publicly exposed AuthorizationError type that returns a PolicyId. I don’t think this is possible without a breaking change.

Agreed. I'm not sure how we would structure it, but the ideal situation would be a public `AuthorizationError` type with a function returning the public `PolicyId` type, but I also don't think that could be done without a breaking change. For now I'd say we drop this function.

khieta replied:
Good catch. No, we don't want to expose a function that returns a `PolicyID`. I'll make a separate PR to change the `AuthorizationError` definition in the next major release.

john-h-kastner-aws commented on cedar-policy/src/api.rs:
```diff
@@ -904,6 +904,11 @@ impl Diagnostics {
     }
+
+    /// Get the `PolicyId`s of the policies where an error occurred during authorization.
+    pub fn error_ids(&self) -> impl Iterator<Item = &PolicyId> {
```
Maybe a more explicit name like `error_policy_ids`. `error_ids` makes it sound like they might be error code.

khieta replied:
Fair point. Changed in the latest commit.

andrewmwells-amazon commented:
> Looks like there's a downstream failure in CedarJava. This should be easy to fix, but does beg the question of whether this is PR is truly backwards compatible. I maintain that it is because our integration test format should not be considered part of the stable API... but open to second opinions.

I think you can safely override this. I still consider it backwards compatible.

andrewmwells-amazon commented:
> The “errors” field of the InterfaceDiagnostics type should by HashSet<PolicyId> instead of HashSet<String>.
Agree we should probably change but don't think it's urgent.

> In the Response structure, the field is called reason (singular), whereas in the integration tests it is called reasons (plural). I didn’t change these to be consistent because I’m conflicted about the right option. I think “reasons” makes more sense since there may be multiple policy ids (and it’s consistent with “errors”), but changing the Response type is a breaking change.
I could see saying "The reason is this (non-empty) list." or "The reasons are the items on this list." I prefer the later, but I'm not convinced it's worth a breaking change.

> Rather than adding an error_ids method to Diagnostics, it would make more sense to add an id method to the publicly exposed AuthorizationError type that returns a PolicyId. I don’t think this is possible without a breaking change.
Confused by this. Adding new methods shouldn't be a breaking change?

john-h-kastner-aws commented:
> Confused by this. Adding new methods shouldn't be a breaking change?

As I understand it, the issue is that `AuthorizationError` is in `cedar-policy-core`, so a function on it can't return `cedar-policy::PolicyId` without moving the enum into `cedar-policy`.

khieta commented:
Thanks for the reviews! I believe I've addressed all comments.

> 2. In the `Response` structure, the field is called `reason` (singular), whereas in the integration tests it is called `reasons` (plural).

I changed the name in the corpus test format because the inconsistency was bothering me. This is not a breaking change.

> 3. Rather than adding an `error_ids` method to `Diagnostics`, it would make more sense to add an `id` method to the publicly exposed `AuthorizationError` type that returns a `PolicyId`. I don’t think this is possible without a breaking change.

Like @john-h-kastner-aws said, the reason this would be breaking is that we'd have to (re)define `AuthorizationError` in `cedar-policy`.

cdisselkoen commented:
> Like @john-h-kastner-aws said, the reason this would be breaking is that we'd have to (re)define `AuthorizationError` in `cedar-policy`.

If `AuthorizationError` is already a publically exported type from `cedar-policy`, why does changing where it's defined constitute a breaking change?

khieta commented:
Fix for Java CI failure: https://github.com/cedar-policy/cedar-java/pull/78

khieta commented:
> If `AuthorizationError` is already a publicly exported type from `cedar-policy`, why does changing where it's defined constitute a breaking change?

`AuthorizationError` is currently defined in `cedar-policy-core` (and needs to stay there because it has other dependencies). There is no method to pull out the `id` field, but you can pattern match to get it. The ideal change (IMO) is to define a new `AuthorizationError` in `cedar-policy` that is a wrapper around the internal version, and not publicly expose the internal version. This means that you could no longer pattern match on the internal error type, which is a change in behavior.

cdisselkoen commented:
I suppose instead of wrapping the internal version, we could mirror the internal version, similar to how `SchemaError` works.  Then users could still pattern match on the internal error type

khieta commented:
> I suppose instead of wrapping the internal version, we could mirror the internal version, similar to how SchemaError works. Then users could still pattern match on the internal error type

I hadn't thought about that. The `error_policy_ids` method felt a little clunky, so I switched to your suggestion in the latest (non-merge) commit. Unfortunately, we have to reference the internal `PolicyID` type in the struct to avoid a breaking change, but at least the `id()` method can return a `PolicyId`. Lmk what you think!

john-h-kastner-aws commented on cedar-policy/src/api.rs:
```diff
@@ -735,6 +734,40 @@ impl Authorizer {
+    #[error("while evaluating policy `{id}`: {error}")]
+    PolicyEvaluationError {
+        /// Id of the policy with an error
+        id: ast::PolicyID,
```
We should `#[doc(hidden)]` this to avoid confusion with the actual getter function. That should be fair game for a minor release since you can still access it if really want to.

khieta replied:
Good call. Done in the last commit.

john-h-kastner-aws reviewed (approved):
Seems legit. I like this better and AFAIK it shouldn't be a breaking change.

khieta commented:
Pushed one more edit -- I just realized that `cedar-policy-cli` was redefining `JsonTest` and `JsonRequest` when it could be importing them from `cedar_policy` instead.

I'm done now I swear!
